### ISSUE TITLE
Hearing cracking/popping, and noticing CSS animation framerate drop after running midio for awhile

### ISSUE THREAD

**Issue by evangipson** [evangipson/midio#26]
## Expected Behavior
You'd never hear popping or crackling in the audio, and never notice a slowdown no matter how long you let midio run.

## Actual Behavior
CRACK *beautiful noises* POPPLE POP POP *gorgeous airy chord* CRAAAAAAckckkckKCkkKKk

## Solutions
1. the ```<span>```s that i'm creating for the piano key effects might not be getting cleaned up correctly? i am using .removeChild and not assigning it to a variable... so it should just go away. Maybe i could only allow X number of spans to be created??
1.  performance of many oscillators being created at a time with many filters each oscillator. i don't think this is the problem because if i load up midio and click a bunch it's fine and doesn't start clicking/crackling/slowing down til after a bunch of clicking. it's like the problem always takes around a minute to show up- and sometimes it never shows up at all.. .which leads me to my third theory
1.  some filter being applied to some oscillator at some volume level sometimes. this might be due to any number of calculations in the code for ADSR, pretty much anytime i'm doing any filter or gain adjustments on any filter or oscillator- which is kind of a lot. i believe this is my strongest lead.

## Steps to Reproduce the Problem
1. Load up midio
1. Click a bunch
1. Notice how you don't get crackling/popping/slow down all of the time
1. Reload midio
1. Let it run for 1-5 minutes
1. See if you hear crackling

## Specifications
  - Midio Version: current
  - Operating System: Windows 10
  - Node Version: v0.10.25

**Comment by robertpateii** [comment]
i think i figured this out and sent you messages (and gifs) on teams about it. i don't want to spoil the ending here

**Comment by robertpateii** [comment]
hint: you're really close with "using .removeChild and not assigning it to a variable... so it should just go away. "

it doesn't just go away. none of the objects you stop using do. They hang around until the garbage collector decides to trace out all the objects and figure out which ones aren't in use then remove them from memory in one big go.

This is why high performance games are not written in garbage collected languages (and if they are, the garbage collector's behavior is tip toed around and carefully managed and understood).

**Comment by evangipson** [comment]
This was addressed in https://github.com/evangipson/midio/pull/29 (more specifically it began with commit https://github.com/evangipson/midio/commit/ad86ac080bbb0708f00d298e4e98546b05c49f7f and ended right after with commit https://github.com/evangipson/midio/commit/c7148d49f6bb836656c4d39e54f1bb245ee2d5d1

**Comment by robertpateii** [comment]
the new look is fantastic

**Comment by evangipson** [comment]
I don't think this is a JavaScript heap problem. I've done [some research on Oscillators](http://blog.szynalski.com/2014/04/web-audio-api/) and seems like you can't .start() an oscillator that you've used .stop() on. The garbage collector will clean up the oscillator after a .stop() is called.

Then I took a few heap snapshots after some heavy usage and they were all the same as the heap when the program loaded... so I don't think I have any leaks on the JS side. I think it might be due to me using "opacity" and an animation on it, forcing a repaint perhaps.

**Comment by evangipson** [comment]
After some more hard testing this morning - I've found out a couple things:
- When I display: none the note-active class... there isn't much performance difference. I'm beginning to think it's the JS again, and not the CSS because my animations are optimized now.
- I'm noticing some slowdown with user input notes after awhile... this could be related to the issue I'm having perhaps? Like you'll click around the on the visualizer and not see your notes appear for 5-10 seconds.

**Comment by robertpateii** [comment]
Or by "it's the js" do you mean purely the music genration? What if you ripped out the visual elemets entirely as a test?

**Comment by evangipson** [comment]
yeah it's the music generation, or something else in the JS i believe, because I turned off everything relating to drawing the notes and was still getting performance issues (crackling/popping/lagging input/fan blowing like crazy in laptop) after awhile, even after disabling the drawNoteOnVisualizer() function , which handles updating the DOM.
